sources/dma_driver.c

```diff
@@ -0,0 +1,719 @@
+    dma_chsel(dma_stream, channel); 
+
+    // Configure the direction 
+    dma_dir(dma_stream, dir); 
+
+    // Configure circular 
+    dma_cm(dma_stream, cm); 
+    
+    // Configure the stream priority 
+    dma_priority(dma_stream, priority); 
+    
+    // Configure increment/fixed memory mode 
+    dma_minc(dma_stream, minc); 
+    dma_pinc(dma_stream, pinc); 
+    
+    // Configure data widths 
+    dma_msize(dma_stream, msize); 
+    dma_psize(dma_stream, psize); 
+} 
+
+
+// Configure the DMA stream 
+void dma_stream_config(
+    DMA_Stream_TypeDef *dma_stream, 
+    uint32_t per_addr, 
+    uint32_t mem_addr, 
+    uint16_t data_items)
+{
+    // Configure the total number of data items to be transferred 
+    dma_ndt(dma_stream, data_items); 
+
+    // Set the peripheral port address 
+    dma_par(dma_stream, per_addr); 
+
+    // Set the memory address and subsequently double buffer mode if needed 
+    dma_m0ar(dma_stream, mem_addr); 
+
+    // Enable the DMA stream 
+    dma_stream_enable(dma_stream); 
+}
+
+
+// Configure the FIFO usage 
+void dma_fifo_config(
+    DMA_Stream_TypeDef *dma_stream, 
+    dma_fifo_mode_t mode, 
+    dma_fifo_threshold_t fth, 
+    dma_feie_t feie)
+{
+    // FIFO mode 
+    dma_dmdis(dma_stream, mode); 
+
+    // FIFO threshold 
+    dma_fth(dma_stream, fth); 
+
+    // FIFO interrupt enable 
+    dma_feie(dma_stream, feie); 
+}
+
+
+// Configure the DMA interrupts 
+void dma_int_config(
+    DMA_Stream_TypeDef *dma_stream,  
+    dma_tcie_t tcie, 
+    dma_htie_t htie, 
+    dma_teie_t teie, 
+    dma_dmeie_t dmeie)
+{
+    // Transfer complete interrupt 
+    dma_tcie(dma_stream, tcie); 
+
+    // Half transfer interrupt 
+    dma_htie(dma_stream, htie); 
+
+    // Transfer error interrupt 
+    dma_teie(dma_stream, teie); 
+
+    // Direct mode error interrupt 
+    dma_dmeie(dma_stream, dmeie); 
+}
+
+//================================================================================
+
+
+//================================================================================
+// DMA interrupt status registers 
+
+// Clear all the stream interrupt flags LISR and HISR registers 
+void dma_clear_int_flags(
+    DMA_TypeDef *dma)
+{
+    dma->LIFCR = ~((uint32_t)CLEAR); 
+    dma->HIFCR = ~((uint32_t)CLEAR); 
+}
+
+
+// Read the stream interrupt flags 
+void dma_int_flags(
+    DMA_TypeDef *dma, 
+    uint32_t lo_streams, 
+    uint32_t hi_streams)
+{
+    // Record all stream interrupt flags 
+    lo_streams = dma->LISR; 
+    hi_streams = dma->HISR; 
+
+    // Clear all flags 
+    dma_clear_int_flags(dma); 
+}
+
+//================================================================================
+
+
+//================================================================================
+// DMA Stream x Configuration Register 
+
+// Stream enable 
+void dma_stream_enable(
+    DMA_Stream_TypeDef *dma_stream)
+{
+    dma_stream->CR |= (SET_BIT << SHIFT_0); 
+}
+
+
+// Stream disable  
```

Comment:
Nit:

I would advise putting the function description comments with the definition vs the declaration for the prototype.

This way the documentation is closer to the code